Conf count

NishantBansal2003 opened the pull request:
## Change Description
Fixes: #9452 
This PR enhances the `PendingChannelsResponse_PendingChannel` structure to include the `ConfirmationUntilConfirmed` field. This new field provides users with the exact number of confirmations required for a pending channel's to transition to an open state.

### Changes:
- **Database Update**:
The channel's `shortchannelID` is now persisted in the database once its funding transaction receives 1 confirmation. This ensures we have a reference point to track progress toward the required confirmation depth.

- **API Enhancement:**
 Introduced the `ConfirmationUntilConfirmed` field in the `PendingChannelsResponse_PendingChannel` response. This field indicates the remaining confirmations needed for the channel to become active.​
  - If the funding transaction is unconfirmed, `ConfirmationUntilConfirmed` defaults to the total required confirmations (`NumConfsRequired`).

## Steps to Test

- **itest:** itest have been added to validate this new functionality.

- **Manual Verification:** To manually test:​
  - Establish a channel between two peers without confirming the funding transaction.​

  - Execute `lncli pendingchannels` to observe the `confirmation_until_confirmed` field, which should display the remaining confirmations needed for the channel to transition to an open state.


## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change is not [insubstantial](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#substantial-contributions-only). Typo fixes are not accepted to fight bot spam.
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#code-documentation-and-commenting) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

MPins reviewed (commented):
Good job :+1:

saubyk commented on lnrpc/lightning.proto:
```diff
@@ -2847,6 +2847,11 @@ message PendingChannelsResponse {
+        // The number of blocks remaining before the pending channel open
+        // transaction is fully confirmed. A value of 0 indicates that the
+        // transaction has reached the required number of confirmations.
+        int32 confirmation_until_confirmed = 7;
```
This variable should be named `confirmations_until_active` instead, since what we are tracking is channel status change from 'pending' to 'active'

saubyk commented on lnrpc/lightning.proto:
```diff
@@ -2847,6 +2847,11 @@ message PendingChannelsResponse {
         int32 funding_expiry_blocks = 3;
+
+        // The number of blocks remaining before the pending channel open
+        // transaction is fully confirmed. A value of 0 indicates that the
```
its really the number of blocks remaining before the channel status is 'active', not really "full confirmed". Important to distinguish between transaction confirmations and channel status.

saubyk reviewed (changes requested):
providing some initial comments. will test further.

MPins commented on funding/manager_test.go:
```diff
@@ -1496,6 +1496,16 @@ func testNormalWorkflow(t *testing.T, chanType *lnwire.ChannelType) {
 	assertErrorNotSent(t, alice.msgChan)
 	assertErrorNotSent(t, bob.msgChan)
 
+	// We send two notifications:
+	// 1. The first adds the SCID to the database, allowing calculation of
+	//    the number of confirmations before the channel is fully opened.
+	// 2. The second marks the channel as open.
+	alice.mockNotifier.oneConfChannel <- &chainntnfs.TxConfirmation{
+		Tx: fundingTx,
+	}
+	bob.mockNotifier.oneConfChannel <- &chainntnfs.TxConfirmation{
+		Tx: fundingTx,
+	}
```
It works, but this doesn't reflect the actual open channel process, right? Have you considered sending the channel confirmation after the transaction confirmation?

NishantBansal2003 replied:
Sorry, I didn't quite understand what you meant. My understanding is that if we wait until the transaction is confirmed (after 3 or 6 blocks, etc.), then by the time we mark the channel as confirmed, it will already be open. Also, since we need the details of the SCID, we have to mark the channel as confirmed (after 1 block) even before the transaction is fully confirmed. Am I thinking in the right direction, or is there anything I should read first to better align my thoughts with your points?

MPins replied:
I meaning that you are sending the confirmation tx event twice to Alice and Bob funding managers. I'm not sure if this is happening during the normal workflow when a channel is opening.

NishantBansal2003 replied:
> I meaning that you are sending the confirmation tx event twice to Alice and Bob funding managers. I'm not sure if this is happening during the normal workflow when a channel is opening.

Ok, so what I understand is that in a normal, real-world workflow, whenever a confirmation tx notification is registered, the notification is sent to all the goroutines that registered for it as soon as the block is mined. However, since this is a unit test, I need to explicitly send the confirmation tx event twice.

> Have you considered sending the channel confirmation after the transaction confirmation?

Since we need the details of the SCID, we have to mark the channel as confirmed (after one block) even before the transaction is fully confirmed.
Also, if I send the goroutine for channel confirmation after the transaction confirmation, I still need to send the confirmation tx event twice because of the select statement:

> 2. If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. 

See: https://go.dev/ref/spec#Select_statements

To avoid flaky tests, I send the confirmation tx event twice.
I hope this is the correct way of handling the process. Please suggest any better method you have in mind.

MPins reviewed (commented):
I took a deeper look at `manager_test` and have a few questions.

saubyk reviewed (commented):
Tested three scenarios:
1. Regular channel, default confirmations required for active is 3. `confirmations_until_active` value in `pendingchannels` response tracks, until the channel is active. Test pass.
2. Wumbo channel, default confirmations required for active is 6. `confirmations_until_active` value in `pendingchannels` response tracks, until the channel is active. Test pass.
3. Updated the node config with `bitcoin.defaultchanconfs=4`. `confirmations_until_active` value in `pendingchannels` response doesn't track. It still starts off with a value of 3 (should start with 4) and channels gets active after 3 confirmations. Test fail.

Don't think that the issue with last scenario has been introduced with this change. But it would be good to investigate further why the setting in config is not having any bearing on the default behavior.

NishantBansal2003 commented on funding/manager_test.go:
```diff
@@ -407,7 +407,7 @@ func createTestFundingManager(t *testing.T, privKey *btcec.PrivateKey,
 	chainNotifier := &mockNotifier{
 		oneConfChannel: make(chan *chainntnfs.TxConfirmation, 1),
-		sixConfChannel: make(chan *chainntnfs.TxConfirmation, 1),
+		sixConfChannel: make(chan *chainntnfs.TxConfirmation, 6),
```
Based on their names, I initially thought it was a typo; however, I'm not sure if it serves a specific purpose. If the naming is indeed intentional, could you explain why both `oneConfChannel` and `sixConfChannel` are the same?

MPins replied:
The make(..., 6) call here is only sizing the buffer, but the test doesn't actually need a buffer of size 6. The  name sixConfChannel refers to the event (i.e., sixth confirmation received), not the buffer size.

NishantBansal2003 replied:
Thanks, understood. I have reverted this change. PTAL

NishantBansal2003 commented:
> 3\. Updated the node config with `bitcoin.defaultchanconfs=4`. `confirmations_until_active` value in `pendingchannels` response doesn't track. It still starts off with a value of 3 (should start with 4) and channels gets active after 3 confirmations. Test fail.

I checked this locally and it works for me. I have also added an itest for this case. Not entirely sure, but it seems like you might have missed what `bitcoin.defaultchanconfs` means is "We'll require any **incoming** channel requests to wait for this many confirmations before we consider the channel active."

So:
- Alice -> Bob (the `confirmations_until_active ` starts off with a value of Bob's `defaultchanconfs`)
- Bob -> Alice (the `confirmations_until_active ` starts off with a value of Alice's `defaultchanconfs`)

saubyk commented:
> > 3. Updated the node config with `bitcoin.defaultchanconfs=4`. `confirmations_until_active` value in `pendingchannels` response doesn't track. It still starts off with a value of 3 (should start with 4) and channels gets active after 3 confirmations. Test fail.
> 
> I checked this locally and it works for me. I have also added an itest for this case. Not entirely sure, but it seems like you might have missed what `bitcoin.defaultchanconfs` means is "We'll require any **incoming** channel requests to wait for this many confirmations before we consider the channel active."
> 
> So:
> 
> * Alice -> Bob (the `confirmations_until_active ` starts off with a value of Bob's `defaultchanconfs`)
> * Bob -> Alice (the `confirmations_until_active ` starts off with a value of Alice's `defaultchanconfs`)

Good clarification. Tested again by updating the peer's setting and was able to get the correct values for `confirmations_until_active`.

saubyk reviewed (approved):
Great work. tAck.
Code review needs approval from two other devs.

cc: @kaloudis would you like to test this pr and provide some feedback?

NishantBansal2003 commented on channeldb/channel_test.go:
```diff
@@ -978,6 +978,81 @@ func TestChannelStateTransition(t *testing.T) {
+		FundingBroadcastHeight, "broadcast height mismatch")
+
+	confirmedScid := lnwire.ShortChannelID{
+		BlockHeight: 5,
```
Done👍

lightninglabs-deploy commented:
@nishantbansal2003, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@nishantbansal2003, remember to re-request review from reviewers when ready

ziggie1984 commented on rpcserver.go:
```diff
@@ -3919,6 +3919,37 @@ func (r *rpcServer) fetchPendingOpenChannels() (pendingOpenChannels, error) {
+		// required.
+		var remainingConfs int32
+		openTxBlockHeight := int32(
+			pendingChan.ShortChannelID.BlockHeight,
```
Add a comment that zeroconf channels never show up here because they are immediately marked as open from the beginning meaning that they will not show up as pending, otherwise this would cause problems here.

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -2847,6 +2847,11 @@ message PendingChannelsResponse {
         int32 funding_expiry_blocks = 3;
+
+        // The number of blocks remaining until the channel status changes from
+        // pending to active. A value of 0 indicates that the channel is now
```
Also include following comments:

1. Active here means that both channel peers consider this channel as open and can start using it. For public channels this does not mean that the channel is broadcasted to the network yet, this only happens after 6 confirmations which is a BOLT specification. (Link the specifc line of the spec here)

2. ZeroConf channels are active from the beginning which means they never show up as pending from the start.

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -2847,6 +2847,11 @@ message PendingChannelsResponse {
+        // The number of blocks remaining until the channel status changes from
+        // pending to active. A value of 0 indicates that the channel is now
+        // active.
+        int32 confirmations_until_active = 7;
```
We should add two fields, the confirmation height and required_number_of_confirmations_until_acitve or something like this.

ziggie1984 commented on rpcserver.go:
```diff
@@ -3919,6 +3919,37 @@ func (r *rpcServer) fetchPendingOpenChannels() (pendingOpenChannels, error) {
+		// remainingConfs is set to the total number of confirmations
+		// required.
+		var remainingConfs int32
+		openTxBlockHeight := int32(
```
this should be its own function

ziggie1984 commented on rpcserver.go:
```diff
@@ -3940,11 +3971,12 @@ func (r *rpcServer) fetchPendingOpenChannels() (pendingOpenChannels, error) {
+			FeePerKw: int64(localCommitment.
+				FeePerKw),
+			FundingExpiryBlocks:      fundingExpiryBlocks,
+			ConfirmationsUntilActive: remainingConfs,
```
we should also add the confirmation height here.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,92 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// Restart both nodes to test that the appropriate state has been
+	// persisted and that both nodes recover gracefully.
+	ht.RestartNode(alice)
+	ht.RestartNode(bob)
```
Also need the ensure connected directive.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,92 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// Ensure that even if a block is mined and the funding transaction
+	// remains unconfirmed, ConfirmationsUntilActive still equals numConfs
+	ht.MineEmptyBlocks(1)
+	require.Equal(ht, numConfs, ht.AssertNumPendingOpenChannels(alice,
+		1)[0].ConfirmationsUntilActive)
```
Nit: Not necessary this line, I mean we are mining an empty block.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,92 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+
+	// ConfirmationsUntilActive field should decrease as each block is
+	// mined until the required number of confirmations is reached. Let's
+	// mine a few empty blocks and verify the value of
```
we should mine normal blocks here, because there is no tx in the mempool anyways ?

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,92 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// mined until the required number of confirmations is reached. Let's
+	// mine a few empty blocks and verify the value of
+	// ConfirmationsUntilActive at each step.
+	for i := int32(1); i < numConfs; i++ {
```
I think for these cases a reverse loop fits better:

// we could also decrease the numConfs immediately when confirming the tx above which is maybe even better.
```
for i := numConfs-1; i > 0; i-- {
		expectedConfirmationsLeft := i
```

ziggie1984 reviewed (changes requested):
Thanks for picking this issue up 👍.

Nice work so far, the direction is the right one, left some comments for you.

NishantBansal2003 commented on funding/manager.go:
```diff
@@ -3012,6 +3012,16 @@ func (f *Manager) waitForFundingWithTimeout(
+	}
 
 	f.wg.Add(1)
 	go f.waitForFundingConfirmation(ch, cancelChan, confChan)
```
But doesn’t `ConfirmationEvent.Updates` only return the number of confirmations remaining? Since we want to use `chainntnfs.TxConfirmation` to retrieve the SCID details for updating the channel, I used this method.

ziggie1984 replied:
Ok, but I still propose using it, and instead adding a new tlv field to the OpenChannelStruct called `Confirmation_Height` and setting it for every channel (normal and zeroconf). Because that#s the only think we are interested in.

ziggie1984 replied:
Then we can also leave `MarkConfirmedScid ` untouched.

ziggie1984 replied:
I also like it more, because I think the ShortChan ID should only be set when the channel is active not when it is still unusable.

NishantBansal2003 replied:
Hey @ziggie1984, apologies for the late response. So I want to check that my understanding is correct and aligns with what we need:

We should add a new field, `confirmationHeight`, to `openChannelTlvData` and `ConfirmationHeight` to `OpenChannel`. Whenever we receive a confirmation, we’ll record its height in the database. I’ve reviewed the code but couldn’t find where we specify/retrieve the values of `openChannelTlvData`—should I update (and rename) `MarkConfirmedScid` so that it stores this `confirmationHeight` value in the database?

We won’t store the SCID until the channel is active, but we still need to call `confNtfn.Confirmed`, since the confirmation height is only available from `chainntnfs.TxConfirmation`. So `handleOpenChanTxConfirmation` should remain—correct?

ziggie1984 replied:
I created an proto: https://github.com/ziggie1984/lnd/commit/b181704a4c0df97c0604b9c56ebb5356fb07b55b

ziggie1984 replied:
So we need to make sure we update the blockheight in case there is a reorg und reset it basically. Reorgs of 1 block are not frequent but they happen.

ziggie1984 commented on rpcserver.go:
```diff
@@ -3860,6 +3860,36 @@ type (
 	waitingCloseChannels []*lnrpc.PendingChannelsResponse_WaitingCloseChannel
 )
 
+// getRemainingConfs calculates how many blocks remain until the funding
+// transaction reaches the required confirmation height. If the funding
+// transaction has received at least one confirmation(i.e. its block height is
+// known), then confirmationHeight is calculated as: funding transaction block
+// height + required number of confirmations - 1 (subtracting 1 because one
+// confirmation is already achieved).
+// remainingConfs is computed as the difference between confirmationHeight and
+// currentHeight, but is never less than zero.
+//
+// If the funding transaction hasn't been confirmed, remainingConfs is set to
+// the total number of confirmations required.
+func getRemainingConfs(pendingChan *channeldb.OpenChannel,
+	currentHeight uint32) uint32 {
+
+	var remainingConfs uint32
+	openTxBlockHeight := pendingChan.ConfirmationHeight
+
+	if openTxBlockHeight > 0 {
+		confirmationHeight := openTxBlockHeight +
+			uint32(pendingChan.NumConfsRequired) - 1
+		remainingConfs = max(0, confirmationHeight-currentHeight)
+	} else {
+		// If the funding transaction is not confirmed yet,then
+		// remainingConfs will always be NumConfsRequired.
+		remainingConfs = uint32(pendingChan.NumConfsRequired)
+	}
+
+	return remainingConfs
+}
+
```
I find this code part lacks a bit readability what about this ?

```
// calcRemainingConfs calculates how many more confirmations are needed for a
// pending channel to be fully confirmed. It takes into account:
// 1. The current blockchain height
// 2. The block height at which the funding transaction was first confirmed
// 3. The total number of confirmations required for the channel.
func calcRemainingConfs(pendingChan *channeldb.OpenChannel,
	currentHeight uint32) uint32 {

	// If the funding transaction hasn't been confirmed yet,
	// we need all the required confirmations
	if pendingChan.ConfirmationHeight == 0 {
		return uint32(pendingChan.NumConfsRequired)
	}

	// Calculate the target height at which the channel will be fully
	// confirmed. The -1 is because the confirmation height of the first
	// confirmation has to be taken into account.
	targetConfirmationHeight := pendingChan.ConfirmationHeight +
		uint32(pendingChan.NumConfsRequired) - 1

	// In case the current height is already past the target, return 0. This
	// should never happen because the channel should already be moved from
	// pending to open state but we handle this case in case of timing
	// issues.
	if currentHeight >= targetConfirmationHeight {
		return 0
	}

	return targetConfirmationHeight - currentHeight
}
```

NishantBansal2003 replied:
Apologies for writing unreadable code. I’ll try to follow good practices and learn from the codebase. I’m sorry for wasting your time, especially after you shared so many code patches.

ziggie1984 commented on rpcserver.go:
```diff
@@ -3940,11 +3979,14 @@ func (r *rpcServer) fetchPendingOpenChannels() (pendingOpenChannels, error) {
+				FeePerKw),
+			FundingExpiryBlocks:      fundingExpiryBlocks,
+			ConfirmationsUntilActive: remainingConfs,
+			ConfirmationHeight: pendingChan.
```
I think when the confHeight is 0 it will be omitted from the response because it is the zero value for it ?

ziggie1984 replied:
can you check whether this is the case ?

NishantBansal2003 replied:
<img width="874" alt="Screenshot 2025-05-11 at 9 55 15 PM" src="https://github.com/user-attachments/assets/892d0ce5-2a84-40df-a26c-9080df5d4f39" />

`ConfirmationHeight ` field is not being omitted from the JSON-RPC response when its value is 0.

NishantBansal2003 replied:
although the `omitempty` tag is present on the field, its value of 0 still appears in the output. I’m not sure why.

ziggie1984 replied:
ahh ok makes sense thanks for verifying this is the reason:

https://github.com/lightningnetwork/lnd/blob/ee25c228e9a3c638286393762bd3da07ac4b45ec/lnrpc/rpc_utils.go#L29

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
 	ht.MineBlocksAndAssertNumTxes(1, 1)
 }
 
+// testConfirmationsUntilActiveOnPending verifies that as a channel status
```
Nit: testConfirmationsUntilActiveOnPending => testPendingChannelConfirmationUntilActive verifies the value for the rpc filed `ConfirmationUntilActive` decreases as expected as soon as blocks are confirmed.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// before it's open, with the current set of defaults, we'll need to
+	// create a new node instance.
+	alice := ht.NewNodeWithCoins("Alice", nil)
+	bob := ht.NewNode("Bob", []string{
```
I don't think you need to start a new node, but you can change it also in the OpenChannel parameters ?

ziggie1984 replied:
ok however I think the comment is a bit misleading, because there are no default nodes anymore. We need to create nodes now.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	)
+
+	// Mine the first block containing the funding transaction, This
+	// confirms the funding transaction but does not change the channel
```
Nit: confirms the funding transaction but the channel should still remain pending ...

ziggie1984 commented on channeldb/channel_test.go:
```diff
@@ -978,6 +978,91 @@ func TestChannelStateTransition(t *testing.T) {
+	require.NoError(t, err, "unable to mark channel's confirmation height")
+
+	// Ensure the channel remains pending after confirmation.
+	require.True(t, pendingChannels[0].IsPending, "channel should remain "+
```
Q: Not sure if this case here makes sense, because the confirmation height is not related to the pending status directly.

ziggie1984 commented on channeldb/channel_test.go:
```diff
@@ -978,6 +978,91 @@ func TestChannelStateTransition(t *testing.T) {
+		"broadcast height mismatch after re-fetching",
+	)
+
+	// Ensure the original channel state's confirmation height is not
+	// updated before refresh.
+	require.NotEqual(
+		t, channelState.ConfirmationHeight,
+		pendingChannels[0].ConfirmationHeight, "original channel "+
+			"state's ConfirmationHeight should not match before "+
+			"refresh",
+	)
```
why are they not matching ?

NishantBansal2003 replied:
Because the in-memory `channelState` has not been updated—and since `pendingChannels` is used to set the `ConfirmationHeight` in the db—the updated value(of `ConfirmationHeight` for `channelState`) is only fetched after `channelState` is refreshed to reflect the newly persisted state in the db.

ziggie1984 replied:
ahh ok makes sense

ziggie1984 commented on channeldb/channel_test.go:
```diff
@@ -978,6 +978,91 @@ func TestChannelStateTransition(t *testing.T) {
+	)
+
+	// Confirm the channel remains pending after refresh.
+	require.True(t, channelState.IsPending, "channel should remain "+
```
See comment above.

ziggie1984 commented on channeldb/channel_test.go:
```diff
@@ -1007,7 +1092,7 @@ func TestFetchPendingChannels(t *testing.T) {
 
 	chanOpenLoc := lnwire.ShortChannelID{
-		BlockHeight: 5,
+		BlockHeight: broadcastHeight + 1,
```
was this a bug ? Wonder why the testcase still passed tho ?

NishantBansal2003 replied:
No, this wasn't a bug. These were only used to check whether the values of scid were persisted in the database. I updated it since the opening transaction was broadcast at broadcastHeight; we could confirm it at broadcastHeight + 1. It was just a nit.

NishantBansal2003 commented on funding/manager.go:
```diff
@@ -3123,34 +3124,110 @@ func (f *Manager) waitForFundingConfirmation(
+			// channel fires before Updates. When multiple cases in
+			// a select are ready, Go makes a uniform pseudo-random
+			// choice between them.
+			if completeChan.ConfirmationHeight == 0 {
```
I was thinking we don’t need this case (though I tried running the code with it and ran into issues because the confirmation notification was sent before anyone was listening, causing a block). We only really need this case when `numConfs` is 1, which leads to two scenarios:

1. The transaction isn’t confirmed yet, so we’ll have `ConfirmationsUntilActive = 1` and `ConfirmationHeight = 0`.
2. The transaction is already confirmed, meaning the channel is no longer pending (it’s active).

In either case, we don’t need to persist the confirmation height in the database. wdyt?

So I’m going to remove the code that marks `ConfirmationHeight` in the database in the `Confirmed` case of that `select`.

ziggie1984 replied:
hmm but we still want our channel struct to have the correct blockheight for this case, so I don't recommend neglecting this case. Normally it should not block because the confirmed is a buffered channel, can you be more specific which test made problems ?

ziggie1984 commented on rpcserver.go:
```diff
@@ -3860,6 +3860,37 @@ type (
+	currentHeight uint32) uint32 {
+
+	// If the funding transaction hasn't been confirmed yet,
+	// we need all the required confirmations
```
Nit: missing point at the end.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	)
+
+	// Since we want Bob's channels to require more than 1 on-chain
+	// confirmation before active, we will launch Bob with the custom
```
Nit: before becoming active ...

ziggie1984 reviewed (commented):
I think we are close 🤝, tho the unit-test failures seem to be related to this change?

NishantBansal2003 commented:
> unit-test failures seem to be related to this change?

Fixed now!

ziggie1984 commented on funding/manager.go:
```diff
@@ -3123,55 +3124,120 @@ func (f *Manager) waitForFundingConfirmation(
 
-	var confDetails *chainntnfs.TxConfirmation
-	var ok bool
+	// Monitor the confirmation event for the funding transaction. And
```
Nit: Monitor block confirmation events ...

ziggie1984 commented on funding/manager.go:
```diff
@@ -3123,55 +3124,120 @@ func (f *Manager) waitForFundingConfirmation(
+			// funding transaction was reorged out.
+			err := completeChan.MarkConfirmationHeight(uint32(0))
+			if err != nil {
+				log.Errorf("failed to update state for "+
```
Nit: instead of a general state use: failed to reset confirmation height for ...

ziggie1984 commented on funding/manager_test.go:
```diff
@@ -178,9 +178,11 @@ func (m *mockAliasMgr) DeleteSixConfs(lnwire.ShortChannelID) error {
 }
 
 type mockNotifier struct {
```
Can you add a reorg-unit-test here, should be easy I think.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
 }
 
+// testPendingChannelConfirmationUntilActive verifies the value for the rpc
+// filed ConfirmationUntilActive decreases as expected as soon as blocks are
```
filed ? did you mean field ? => rpc field `ConfirmationUntilActive` updates correctly as soon as ..

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// Since the funding transaction is not confirmed yet,
+	// ConfirmationsUntilActive will always be numConfs, and confirmation
+	// height will be 0.
+	require.Equal(ht, numConfs, ht.AssertNumPendingOpenChannels(alice,
+		1)[0].ConfirmationsUntilActive)
+	require.Equal(ht, numConfs, ht.AssertNumPendingOpenChannels(bob, 1)[0].
```
Nit: Compare first all the values of Alice then do a linebreak and compare the values of bob.

ziggie1984 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,114 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// Mine the first block containing the funding transaction, This
+	// confirms the funding transaction but the channel should still remain
+	// pending.
+	ht.MineBlocksAndAssertNumTxes(1, 1)
```
I think we are fine for potential slow updates of the Confirmation Height in our backend because we restart the node, otherwise we would have to wait here to make sure we do not progress until the confHeight is set in our db.

lightninglabs-deploy commented:
@mpins: review reminder
@ziggie1984: review reminder

bitromortac commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,115 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	param := lntest.OpenChannelParams{Amt: chanAmt}
+	ht.OpenChannelAssertPending(alice, bob, param)
+
+	// Both Alice and Bob have one pending open channel.
+	ht.AssertNumPendingOpenChannels(alice, 1)
+	ht.AssertNumPendingOpenChannels(bob, 1)
+
+	// Since the funding transaction is not confirmed yet,
+	// ConfirmationsUntilActive will always be numConfs, and confirmation
+	// height will be 0.
+	require.Equal(ht, numConfs, ht.AssertNumPendingOpenChannels(alice,
+		1)[0].ConfirmationsUntilActive)
+	require.Equal(
+		ht, uint32(0),
+		ht.AssertNumPendingOpenChannels(alice, 1)[0].ConfirmationHeight,
+	)
+
+	require.Equal(ht, numConfs, ht.AssertNumPendingOpenChannels(bob, 1)[0].
+		ConfirmationsUntilActive)
+	require.Equal(
+		ht, uint32(0),
+		ht.AssertNumPendingOpenChannels(bob, 1)[0].ConfirmationHeight,
+	)
```
It would be more readable this way, also later:
```golang
	// Both Alice and Bob have one pending open channel.
	alicePending := ht.AssertNumPendingOpenChannels(alice, 1)[0]
	bobPending := ht.AssertNumPendingOpenChannels(bob, 1)[0]

	// Since the funding transaction is not confirmed yet,
	// ConfirmationsUntilActive will always be numConfs, and confirmation
	// height will be 0.
	require.Equal(ht, numConfs, alicePending.ConfirmationsUntilActive)
	require.EqualValues(ht, 0, alicePending.ConfirmationHeight)

	require.Equal(ht, numConfs, bobPending.ConfirmationsUntilActive)
	require.EqualValues(ht, 0, bobPending.ConfirmationHeight)
```

All this may be flaky later, so maybe extract it into a helper function with a `wait.NoError` to wait for the expected values to appear?

bitromortac reviewed (commented):
Thank you for your work and the tests :pray:. Those could be improved a bit in terms of readability, but they look well thought out. I also left a comment about the marking of the confirmation height, I'll need to give this another thought.

NishantBansal2003 commented on itest/lnd_open_channel_test.go:
```diff
@@ -877,6 +877,115 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	ht.RestartNode(alice)
+	ht.RestartNode(bob)
+	ht.EnsureConnected(alice, bob)
+
```
Sure, let’s also check this before restarting, since we’re using `wait.NoError`

bitromortac replied:
Ah yes, you're right, though we'd mine a block first.

NishantBansal2003 commented on funding/manager_test.go:
```diff
@@ -1453,6 +1494,162 @@ func assertHandleChannelReady(t *testing.T, alice, bob *testNode,
+	// Check that the confirmation height is set to 3 for both alice and
+	// bob.
+	assertConfirmationHeight(t, alice, chanID, 3)
+	assertConfirmationHeight(t, bob, chanID, 3)
```
I don’t think so. I just added it to complete the tx workflow; however, if it’s unnecessary, let’s remove it from here

NishantBansal2003 commented on chainntnfs/txnotifier_test.go:
```diff
@@ -271,12 +271,14 @@ func TestTxNotifierFutureConfDispatch(t *testing.T) {
+			NumConfsLeft: 0,
+			BlockHeight:  11,
+		}
+		if !updDetails.Equal(&expected) {
```
Makes sense. Should I also remove the convenience methods from `TxUpdateInfo`?

bitromortac commented on funding/manager.go:
```diff
@@ -3123,55 +3124,120 @@ func (f *Manager) waitForFundingConfirmation(
+// handleFinalConfirmation is a helper function that listens for the final
+// confirmation event and sends the final confirmation to the confirmation
+// channel which was provided by the caller.
+func (f *Manager) handleFinalConfirmation(
```
nice, thank you for the speedy turnaround @NishantBansal2003 :pray:! I'll review it after next week, since I'll be out

lightninglabs-deploy commented:
@mpins: review reminder
@ziggie1984: review reminder
@nishantbansal2003, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mpins: review reminder
@ziggie1984: review reminder
@nishantbansal2003, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mpins: review reminder
@ziggie1984: review reminder
@nishantbansal2003, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mpins: review reminder
@ziggie1984: review reminder
@nishantbansal2003, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mpins: review reminder
@ziggie1984: review reminder
@nishantbansal2003, remember to re-request review from reviewers when ready

bitromortac commented on funding/manager_test.go:
```diff
@@ -1545,13 +1665,28 @@ func testNormalWorkflow(t *testing.T, chanType *lnwire.ChannelType) {
+	// here.
+	alice.mockNotifier.sixUpdateChannel <- chainntnfs.TxUpdateInfo{
+		NumConfsLeft: 0,
+		BlockHeight:  1,
+	}
+	bob.mockNotifier.sixUpdateChannel <- chainntnfs.TxUpdateInfo{
+		NumConfsLeft: 0,
+		BlockHeight:  1,
+	}
+
 	// Notify that six confirmations has been reached on funding
 	// transaction.
 	alice.mockNotifier.sixConfChannel <- &chainntnfs.TxConfirmation{
-		Tx: fundingTx,
+		Tx:          fundingTx,
+		BlockHeight: 1,
 	}
 	bob.mockNotifier.sixConfChannel <- &chainntnfs.TxConfirmation{
-		Tx: fundingTx,
+		Tx:          fundingTx,
+		BlockHeight: 1,
```
Ah right, sorry, got confused here

ziggie1984 reviewed (commented):
Thank you for finalizing this PR, had some final nits.

ziggie1984 commented on funding/manager.go:
```diff
@@ -3108,34 +3109,111 @@ func (f *Manager) waitForFundingConfirmation(
+
+			return
+		}
 	}
```
@gemini-code-assist  I find this return value in case of an error even more confusing when reading the code, I actually think we don't need this helper, it makes the code more readable without, what do you think.

ziggie1984 replied:
@NishantBansal2003 can you revert this change pls!

ziggie1984 commented on funding/manager.go:
```diff
@@ -3108,34 +3109,130 @@ func (f *Manager) waitForFundingConfirmation(
-			completeChan.FundingOutpoint)
-		return
-	}
+			log.Debugf("funding tx %s received confirmation in "+
```
can you here in this log line also mention how many confirmations we still need to wait `NumConsLeft` otherwise it can be misleading.

ziggie1984 commented:
> should I add an itest for this?

I don't think it is super important for this feature, however if you are interested you can look at `testOpenChannelAfterReorg` how to design a reorg test.

NishantBansal2003 commented:
Oops, apologies, everything is done locally, only my testing remains. I will test and push by the end of today (IST timezone)

ziggie1984 commented on chainntnfs/txnotifier.go:
```diff
@@ -2011,6 +2011,20 @@ func (n *TxNotifier) dispatchConfReorg(ntfn *ConfNtfn,
+
+		// We also signal the reorg to the notifier in case the
+		// subscriber is also interested in the reorgs before the
+		// transaction received it required confirmation.
```
it => its required confirmations.

ziggie1984 reviewed (approved):
Pending CI, otherwise excellent work, LGTM  🎉
